client/state/products-list/actions.js

```diff
@@ -17,19 +12,33 @@ import {
 import 'calypso/state/products-list/init';
 
 export function receiveProductsList( productsList ) {
+	// Since the request succeeded, productsList should be guaranteed non-null;
+	// thus, we don't have any safety checks before this line.
+	Object.keys( productsList ).forEach(
+		( slug ) => ( productsList[ slug ] = createProductObject( productsList[ slug ] ) )
```

Comment:
This looks a bit odd for a couple of reasons:

* We're mutating `productsList` which is the argument that is passed to `receiveProductsList`. Do you think we could create a copy instead? Essentially, replacing `forEach` with a `map()` and using the resulting new array for `productsList` should manage that.
* On this line we're both assigning the value to `productsList[ slug ]` and returning it (but the returned value isn't used). If the returned value isn't used, perhaps we shouldn't return at all?

Reply:
I think it's definitely reasonable for us to avoid mutating the `productsList` argument. Honestly it was an oversight on my part, as I thought that `mapValues` was also mutating. (Testing in the console shows that it doesn't.) Will address! 👍